validator/nodejs/index.js

```diff
@@ -142,6 +147,100 @@ function readFromUrl(url, userAgent) {
 }
 
 /**
+ * Creates a promise returning a list of crawled URLs beginning at the given
+ * start URL. Supported options are:
+ *
+ * <pre><code>
+ * {
+ *   crawlMaxDepth: 1,
+ *   crawlMaxConcurrency: 5,
+ *   userAgent: 'amphtml-validator'
+ * }
+ * </code></pre>
+ * @param {!string} startUrl
+ * @param {!Object} program
+ * @returns {!Promise<!string[]>}
+ */
+function crawlUrl(startUrl, program) {
+  return new Promise(function(resolve, reject) {
+    // Create and configure the crawler.
+    var crawler = new SimpleCrawler(startUrl);
+    crawler.maxDepth = program.crawlMaxDepth;
+    crawler.userAgent = program.userAgent;
+    crawler.maxConcurrency = program.crawlMaxConcurrency;
+
+    var discoveredAmpDocs = [];
+
+    // Queue AMPs for validation and discover outgoing links
+    crawler.discoverResources = function(buffer, queueItem) {
+      var body = buffer.toString('utf8');
+      var $ = cheerio.load(body);
```

Comment:
It's a little sad that this introduces a second html parser, but I see that it's attractive for doing dom queries and such, and commonly used with the SimpleCrawler library.